sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/AttributesProcessor.java

```diff
@@ -0,0 +1,198 @@
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.context.Context;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.function.BiFunction;
+import java.util.function.Predicate;
+import java.util.function.UnaryOperator;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * An {@code AttributesProcessor} is used by {@code View}s to define the actual recorded set of
+ * attributes.
+ *
+ * <p>An AttributesProcessor is used to define the actual set of attributes that will be used in a
+ * Metric vs. the inbound set of attributes from a measurement.
+ */
+@Immutable
+public abstract class AttributesProcessor {
+  private AttributesProcessor() {}
+
+  /**
+   * Manipulates a set of attributes, returning the desired set.
+   *
+   * @param incoming Attributes associated with an incoming measurement.
+   * @param context The context associated with the measurement.
+   */
+  public abstract Attributes process(Attributes incoming, Context context);
+
+  /**
+   * If true, this ensures the `Context` argument of the attributes processor is always accurate.
+   * This will prevents bound instruments from pre-locking their metric-attributes and defer until
+   * context is available.
+   */
+  public abstract boolean usesContext();
+
+  /** Joins this attribute processor with another that operates after this one. */
+  public AttributesProcessor then(AttributesProcessor other) {
+    if (other == NOOP) {
+      return this;
+    }
+    if (this == NOOP) {
+      return other;
+    }
+
+    if (other instanceof JoinedAttribtuesProcessor) {
+      return ((JoinedAttribtuesProcessor) other).prepend(this);
+    }
+    return new JoinedAttribtuesProcessor(Arrays.asList(this, other));
+  }
+
+  /** No-op version of attributes processer, returns what it gets. */
+  public static AttributesProcessor noop() {
+    return NOOP;
+  }
+
+  /**
+   * Creates a processor which filters down attributes from a measurement.
+   *
+   * @param nameFilter a filter for which attribute keys to preserve.
+   */
+  @SuppressWarnings("unchecked")
+  public static AttributesProcessor filterByKeyName(Predicate<String> nameFilter) {
+    return simple(
+        incoming -> {
+          AttributesBuilder result = Attributes.builder();
+          incoming.forEach(
+              (k, v) -> {
+                if (nameFilter.test(k.getKey())) {
+                  result.put((AttributeKey<Object>) k, v);
+                }
+              });
+          return result.build();
+        });
+  }
+
+  /**
+   * Creates a processor which appends values from {@link Baggage}.
+   *
+   * <p>These attributes will not override those attributes provided by instrumentation.
+   *
+   * @param nameFilter a filter for which baggage keys to select.
+   */
+  public static AttributesProcessor appendBaggageByKeyName(Predicate<String> nameFilter) {
```

Comment:
Hmm - could we leave out baggage from this initial PR? I think it can use more thought.

My initial thought is this isn't where to map baggage to attributes, that should probably happen in the instrumentation, who's "job" is to create the attributes for traces and metrics. At least in the way we've defined it so far.

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java#L140

That way, both traces and metrics will get the baggage. And metrics an use the normal filtering mechanism for attributes without a separate baggage-specific one. I suspect any approach we take to baggage we'd want to map into attributes, independently of signal, and let the signals consume them.

Reply:
I hear what you're saying, but that discussion (likely) belongs in the specification sig.  Baggage is part of the View SDK specification right now:

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#view

> The extra dimensions which come from Baggage/Context (optional). If not provided, no extra dimension will be used. Please note that this only applies to synchronous Instruments.


In the interest of not letting perfect be the enemy of good, I think the current implementation is ok for a few reasons:

1. We've hidden (on purpose) Attributes processor in `internal` so we can evolve it going forward.
2. Baggage appending is unit tested to only be additive.  If instrumentation pre-adds baggage we won't overwrite those labels.
3. Most of the instrumentation work you have regarding baggage is unspecified (AFAIK), meanwhile we have a clear use case and interface via Views.  I suspect we can (and should) find a way to complement the two going forward, but it doesn't have to be now.  This is similar to how each signal needs to independently provide useful feature and then *also* work well when combined in the instrumenter interface.
4. When it comes to leveraging baggage-in-traces (or logs), I actually feel there's a missing component in the SDK that should be doing this, especially for the non-instrumentation cases.

Reply:
The spec repo is a wall of MD and either no comments, or too many comments from a variety of biased vendors - I find discussions there very unproductive. So I'm hoping to take an approach of leaving feedback here and perhaps you will take it to other discussions, or not. I don't intend to block anything, just providing information / suggestions which may be useful. I know it's not fair for me to expect big changes without actually going through the grunt work of the spec - so it's fine if changes aren't made.

I think my biggest reaction here was that we have a path from baggage to metrics already, but even though there's demand we don't have it for traces, which is launched. It's surprising - but it's OK :-)

Reply:
A great point: how is it that we are building this into metrics before we build it into spans?